[SE-0299][TypeChecker] Allow static member references on protocol metatypes in generic contexts

xedin opened the pull request:
Allow referencing static members declared on a protocol or in its extension via
a protocol metatype, but only if extension or member itself bind `Self` of a protocol
to some concrete type via same-type generic requirement.

This requirement allows us to implement a feature via replacing protocol metatype 
with a conforming type as a base for any member reference chain without having to 
search for conforming types (which isn't feasible).

Implementation covers implicit references only e.g.

```swift
protocol P {}
struct S: P {}

extension P where Self == S {
  static var bar: Self { S() }
}

func foo<T: P>(_: T) {
}

foo(.bar)
```

Here base is inferred as `S.Type` based on the fact that `extension S` has `Self` bound to
conforming type `S`. 

Resolves: rdar://problem/43615577

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Allow static member references on protocol metatypes

xedin committed:
[CSBindings] Allow base type of an unresolved member to be inferred from protocol requirements

To be able to infer base type in situations like:

```swift
func foo<T: P>(_: T) {}
foo(.bar)
```

Type variable used for a base type of the chain has to be allowed
to infer types from contextual protocol requirements e.g. `T: P`
in our example.

xedin committed:
[CSSimplify] Change equality semantics between base and result of unresolved member ref

Since it's now possible to infer base type of an unresolved member
reference to be a protocol type we need to adjust `matchTypes` to
check conformance instead of type equality between base (represented
by a protocol) and chain's concrete result type.

xedin committed:
[ConstraintSystem] Adjust handling of incorrect member references on protocol metatypes

Since it's now possible to refer to static members declared on a protocol
metatype if result type conforms to the protocol we need to adjust failure
detection to identify that conformance failure means and invalid reference
in certain situations.

xedin committed:
[CSBindings] Don't infer protocol types from the base of an unresolved member chain

xedin committed:
[CSFix] Add a new fix to diagnose invalid static member refs on protocol metatype

xedin committed:
[Diagnostics] Implement a tailored diagnostic for invalid static member refs on protocol metatypes

xedin committed:
[ConstraintSystem] Add a new constraint which connects base with result of the member chain

The first type represents a result of an unresolved member chain,
and the second type is its base type. This constraint acts almost
like `Equal` but also enforces following semantics:

- It's possible to infer a base from a result type by looking through
  this constraint, but it's only solved when both types are bound.

- If base is a protocol metatype, this constraint becomes a conformance
  check instead of an equality.

xedin committed:
[CSGen] Use new `UnresolvedMemberChainBase` constraint

xedin committed:
[Diagnostics] Extract invalid result type from overload choice

Instead of threading it through the constraint system all the way
to a diagnostic, let's extract it from overload choice instead.

xedin committed:
[ConstraintSystem] Add a new reason for invalid member reference

This new reason indicates that result type of a static member
doesn't conform to a specific protocol so it can't be referenced
on a protocol metatype.

xedin committed:
[ConstraintSystem] Filtering for static member lookup on protocol metatypes

Detect and filter some of the overload choices which can't possibly
match because their result type cannot conform to the expected protocol.

xedin committed:
[ConstraintSystem] Diagnose conformance failure with base of a static member lookup

Produce a tailored diagnostic when it has been established that
base type of a static member reference on protocol metatype doesn't
conform to a required protocol.

xedin committed:
[ConstraintSystem] Mark member result type as a potential hole if base was incorrect

If it has been established that member found via static member
lookup on protocol metatype doesn't have correct result type,
chain's result type the should be treated as a hole.

xedin committed:
[Diagnostics] Add a clarifying note about result type for invalid member ref diagnostic

xedin committed:
[Diagnostics] NFC: Adjust all diagnostics improved/changed by static member refs on protocol metatypes feature

xedin committed:
[TypeChecker] Add tests for static member refs on protocol metatypes feature

xedin committed:
[ConstraintSystem] NFC: Slightly clarify comment about existence static members on a protocol metatype

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please build toolchain

LucianoPAlmeida commented on include/swift/Sema/ConstraintSystem.h:
```diff
@@ -4559,6 +4563,12 @@ class ConstraintSystem {
                                         ConstraintLocatorBuilder locator);
 
+  /// Simplify an equality constraint between result and base types of
+  /// a unresolved member chain.
```
Typo `a unresolved` -> `an unresolved`?

LucianoPAlmeida commented on test/Constraints/static_members_on_protocol_metatype.swift:
```diff
@@ -0,0 +1,215 @@
+_ = P.invalidProp.other
+// expected-error@-1 {{cannot reference static property 'invalidProp' on 'P.Protocol' with non-conforming result type 'Int'}}
+// expected-error@-2 {{value of type 'Int' has no member 'other'}}
+_ = P.invalidMethod()
```
@xedin Just for curiosity, is possible for one to try reference the method like `_ = P.invalidMethod`? If it is what happens in this case?

xedin replied:
It's possible to get a function by value if its result type conform to a required protocol because it would just mean partial application, looks like I didn't add any tests for that, although I should have...

LucianoPAlmeida replied:
>  it would just mean partial application

Ah got it, thanks :)

xedin committed:
[CSFix] NFC: Fix a bad rebase of invalid static member lookup fix

xedin committed:
[TypeChecker] NFC: Add a couple of test-cases of partially applied static members with protocol metatype base

xedin committed:
[CSApply] Simplify handling of static member refs on protocol metatype

xedin committed:
[ConstraintSystem] Ignore optionals when looking for a concrete base of a static member ref on protocol metatype

xedin committed:
[TypeChecker] NFC: Adjust test-case that now prints a generic parameter instead of a hole

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please build toolchain macOS

xedin commented:
@swift-ci please build toolchain macOS

xedin commented:
@swift-ci please build toolchain macOS

xedin committed:
[ConstraintSystem] Limit static member reference on protocol metatype to leading dot syntax only

Do not allow references to a static members with explicitly specified
protocol metatype base, limit that to leading dot syntax only.

xedin commented:
@swift-ci please test

xedin committed:
[CSBindings] Inferring protocol type for member chain base makes binding potentially incomplete

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please smoke test macOS platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please smoke test

xedin committed:
[CSBindings] Update static member lookup to utilize new `isDelayed` API

xedin commented:
@swift-ci please build toolchain macOS

xedin committed:
[CSBindings] NFC: Adjust static member lookup feature to recent `PotentialBindings` changes

- `ConstraintSystem` is now referenced as a member of `PotentialBindings`;
- Literals and defaults are no longer added to the `Bindings` list, so we
  to add a new method `hasViableBindings` to make sure that protocol types
  are added only when there are no other bindings.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[ConstraintSystem] Implement new rule for static member lookup in generic contexts

Instead of requiring result type of the member to conform to declaring protocol,
as originally proposed by SE-0299, let's instead require `Self` to be bound to
some concrete type in extension that declares a static member.

This would make sure that:

1. Members are only visible on a particular type;
2. There is no ambiguity regarding what base of the member chain is going to be.

xedin committed:
[CSDiagnostics] Adjacent diagnostics for the new rules of static member lookup in generic context

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please test Linux platform

xedin committed:
[Diagnostics] Add a tailored diagnostic for contextual lookup on protocols without Self requirement

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please build toolchain macOS

xedin commented:
@swift-ci please build toolchain macOS

xedin commented:
Hopefully this was the last rebase :)

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[CSSimplify] NFC: Adjust to API change - `isHole` -> `isPlaceholder`

hborla commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1902,6 +1902,12 @@ ERROR(type_does_not_conform_owner,none,
       "referencing %0 %1 on %2 requires that %3 conform to %4",
       (DescriptiveDeclKind, DeclName, Type, Type, Type))
+ERROR(contextual_member_ref_on_protocol_requires_self_requirement,none,
+      "contextual member reference to %0 %1 requires "
```
It might be better to use the term 'implicit member reference' because I think that's what they're officially called in TSPL (no need to change this now - we can always bikeshed diagnostics in a follow-up PR)

hborla reviewed (approved):
This looks great to me. Just one small comment about error message wording, but no need to address that now

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility debug

xedin commented:
@swift-ci please test macOS platform
